src/construct.jl

```diff
@@ -348,152 +348,201 @@ function prepare_frame_caller(caller::Frame, framecode::FrameCode, args::Vector{
         else
             if (id = Base.identify_package(mod, String(newname))) !== nothing && haskey(Base.loaded_modules, id)
-                newmod = Base.root_module(id)
+                mod = Base.root_module(id)
             else
-                newmod = Core.eval(mod, :(module $newname end))
+                loc = firstline(ex)
+                mod = Core.eval(mod, Expr(:module, ex.args[1], ex.args[2], Expr(:block, loc, loc)))
             end
         end
-        split_expressions!(modexs, docexprs, lex, newmod, ex.args[3]; extract_docexprs=extract_docexprs, filename=filename)
-    elseif extract_docexprs && is_doc_expr(ex) && length(ex.args) >= 4
-        body = ex.args[4]
-        if isa(body, Expr) && body.head != :call
-            split_expressions!(modexs, docexprs, lex, mod, body; extract_docexprs=extract_docexprs, filename=filename)
-        end
-        docexs = get(docexprs, mod, nothing)
-        if docexs === nothing
-            docexs = docexprs[mod] = Expr[]
-        end
-        if isexpr(body, :module)
-            # If it's a module expression, don't include the entire expression, just document the module itself.
-            excopy = Expr(ex.head, ex.args[1], ex.args[2], ex.args[3])
-            push!(excopy.args, body.args[2])
-            if length(ex.args) > 4
-                append!(excopy.args, ex.args[5:end])   # there should only be a 5th, but just for robustness
+        # We've handled the module declaration, remove it and queue the body
+        pop!(iter.stack)
+        ex = ex.args[3]::Expr
+        push_modex!(iter, mod, ex)
+        return queuenext!(iter)
+    elseif VERSION < v"1.2" && is_function_def(ex)
+        # Newer Julia versions insert a LineNumberNode between statements, but at least Julia 1.0 does not.
+        # Scan the function body for a LNN
+        lnn = firstline(ex)
+        if isa(lnn, LineNumberNode)
+            if iter.lnn === nothing || iter.lnn.line < lnn.line
+                iter.lnn = LineNumberNode(lnn.line-1, lnn.file)
             end
-            push!(docexs, excopy)
-        else
-            push!(docexs, ex)
         end
-    else
-        if isempty(lex.args) || (isexpr(ex, :global) && all(item->isa(item, LineNumberNode), lex.args))
-            push!(modexs, (mod, copy(ex)))
-        else
-            push!(lex.args, ex)
-            push!(modexs, (mod, copy(lex)))
-            empty!(lex.args)
+    elseif head === :macrocall
+        iter.lnn = ex.args[2]
+    elseif head === :block || head === :toplevel
+        # Container expression
+        idx = iter.index[end]
+        while idx <= length(ex.args)
+            a = ex.args[idx]
+            if isa(a, LineNumberNode)
+                iter.lnn = a
+            elseif isa(a, Expr)
+                iter.index[end] = idx + 1
+                push_modex!(iter, mod, a)
+                return queuenext!(iter)
+            else
+                # error("unhandled argument ", a)
```

Comment:
Should this be uncommented or removed?

Reply:
It could be useful for future debugging, but I will remove it. Frequently when I develop code I first write a brittle implementation and then remove the error condition when I become convinced I've handled the necessary cases. But this will throw on many things, so there is little point any more.

Reply:
> Frequently when I develop code I first write a brittle implementation and then remove the error condition when I become convinced I've handled the necessary cases. 

Yeah me too, but without any comment or such it might become a point of uncertainty for a future reader.